# Citation Function Analysis

This repository contains code and data supporting the conclusions made in Byers et. al (CITATION).

## Requirements / Setup

### Install
Example setup using Anaconda:

Create your environment. 

`conda create -n myenv`

Activate your environment. 

`conda activate myenv`

Install pip if not already present.

`conda install pip`

Install required packages using *requirements.txt*.

`pip install -r requirements.txt`

Check installed packages if necessary.

`conda list`

### Environment Variables

This script assumes that two environment variables have been set:
* HF_HOME - local directory location storing your HuggingFace Hub cache and token data.
* HF_TOKEN - Hugging Face Hub User Access Token.

## Analysis Workflow

### Text Generation

The control function `run.py` initiates an iterative batch analysis for a set of data citations. Parameters for a given run are set via JSON configuration files stored in the `config` directory:
* `run_name` - string - Name of batch. Determines name of results sub-directory.
* `four_bit_quant` - boolean - Determines whether or not to quantize the chosen model to four-bit precision. If both `four_bit_quant` and `eight_bit_quant` are set to 'True', the script will choose four-bit precision.
* `eight_bit_quant` - boolean - Determines whether or not to quantize the chosen model to eight-bit precision. If both `four_bit_quant` and `eight_bit_quant` are set to 'True', the script will choose four-bit precision.
* `system_prompt` - string - Determines system prompt supplied to the model for all passes in the batch.
*  `key_pair_dir` - string - Directory path location of .csv file containing the paper>accession pairs constituting the current batch (i.e *"./data/key_pairs/training_pairs.csv"*)
*  `key_meta_dir` - string - Directory path location of .csv file containing the accession metadata for the accessions in the current batch (i.e *"./data/key_metadata/prompt_endpoint_static_INITIAL.csv"*)
*  `do_sample` - boolean - Determines whether or not to allow sampling for certain models.
*  `top_p` - int or null - Determines 'top p' value for certain models.
*  `temperature` - float - Determines temperature value for certain models.

Most of the text generation steps are accomplished using functions from `utils.py`.

Example usage:

`python run.py -c config/config_init0.json`

**Note that this script assumes that your environment has access to adequate GPU and CPU resources. Please check relevant specifications for a given model before running.**

### Outputs

The script generates three output files for each article>accession pair listed in the file specified by the config `key_pair_dir` value:
* RESULTS - Contains only the machine-generated values for 'Data Accessed', 'Use Cases', and 'Tools & Software' for a given article>accession pair.
* MESSAGE - Contains the final formatted chat conversation leading to the values in the RESULTS file.
* PARAMS - Contains each iterative message provided to the model during a chat session along with all selected parameters for each.

## Manual Evaluation

Once a batch is complete, manual evaluation is accomplished using a tabular file generated using `eval_document.py`. This script produces a .csv file containing all of the manual and machine-generated annotations for each article>accession pair. 

Values are assigned in the *score* column as follows:
* False Positive machine-generated value: -1
* True Positive machine-generated value: 1
* False Negative human-generated value: -1
* All others, including redundant values: 0

Refer to Byers, et al (CITATION) for specifics on aggregating redundant values.

Example usage:

`python eval_document.py -o data/results/val_0temp_meta_full_l405 -l data/labels/eval`

Parameters:
* `-o --outdir` : Directory containing output files from `run.py`. It should be *'data/results/{run_name}'*, with *'run_name'* set by your config file.
* `-l --labeldir` : Directory containing manual evaluation JSON files.

This script assumes that you have produced manual labels for all article>accession pairs. See existing examples in *'data/labels'* for formatting specifics.

## Scoring

Precision, Recall, and F1 scores can be generated from your evaluation file using `scoring.py`. This script assumes that all *'score'* values in your evaluation document are set to 0, 1, or -1.

Example usage:

`python scoring.py -e data/results/val_0temp_meta_full_l405/eval1_val_0temp_meta_full_l40.csv`

Parameters:
* `-e --evalpath` : Path to the manual evaluation .csv file generated by `eval_document.py`.

## References

References for all articles included in this repository and in the reported analyses can be seen below.

1. Aati, H. Y., Perveen, S., Aati, S., Orfali, R., Alqahtani, J. H., Al-Taweel, A. M., Wanner, J., & Aati, A. Y. (2022). Headspace solid-phase microextraction method for extracting volatile constituents from the different parts of Saudi Anethum graveolens L. and their antimicrobial activity. Heliyon, 8(3), e09051. https://doi.org/10.1016/j.heliyon.2022.e09051
2. Alex, A., & Antunes, A. (2019). Comparative Genomics Reveals Metabolic Specificity of Endozoicomonas Isolated from a Marine Sponge and the Genomic Repertoire for Host-Bacteria Symbioses. Microorganisms, 7(12), 635. https://doi.org/10.3390/microorganisms7120635
3. Bansal, P., Banda, E. C., Glatt-Deeley, H. R., Stoddard, C. E., Linsley, J. W., Arora, N., Deleschaux, C., Ahern, D. T., Kondaveeti, Y., Massey, R. E., Nicouleau, M., Wang, S., Sabariego-Navarro, M., Dierssen, M., Finkbeiner, S., & Pinter, S. F. (2024). A dynamic in vitro model of Down syndrome neurogenesis with trisomy 21 gene dosage correction. Science Advances, 10(23), eadj0385. https://doi.org/10.1126/sciadv.adj0385
4. Branger, M., Hauer, A., Michelet, L., Karoui, C., Cochard, T., De Cruz, K., Henault, S., Boschiroli, M. L., & Biet, F. (2016). Draft Genome Sequence of Mycobacterium bovis Strain D-10-02315 Isolated from Wild Boar. Genome Announcements, 4(6), e01268-16. https://doi.org/10.1128/genomea.01268-16
5. Burrichter, A. G., Dörr, S., Bergmann, P., Haiß, S., Keller, A., Fournier, C., Franchini, P., Isono, E., & Schleheck, D. (2021). Bacterial microcompartments for isethionate desulfonation in the taurine-degrading human-gut bacterium Bilophila wadsworthia. BMC Microbiology, 21(1), 340. https://doi.org/10.1186/s12866-021-02386-w
6. Chen, H.-X., Chen, F.-J., Zhou, Q.-J., Shang, S.-L., Tang, B., Xu, Z.-J., Duan, L.-J., Jin, J.-L., Xu, G.-Z., Yan, M.-C., & Chen, J. (2024). Two colistin resistance-producing Aeromonas strains, isolated from coastal waters in Zhejiang, China: Characteristics, multi-drug resistance and pathogenicity. Frontiers in Microbiology, 15, 1401802. https://doi.org/10.3389/fmicb.2024.1401802
7. Chorlton, S. D. (2024). Ten common issues with reference sequence databases and how to mitigate them. Frontiers in Bioinformatics, 4, 1278228. https://doi.org/10.3389/fbinf.2024.1278228
8. Chotoo, C. K., Silverman, G. A., Devor, D. C., & Luke, C. J. (2013). A Small Conductance Calcium-Activated K+ Channel in C. elegans, KCNL-2, Plays a Role in the Regulation of the Rate of Egg-Laying. PLoS ONE, 8(9), e75869. https://doi.org/10.1371/journal.pone.0075869
9. Clabaut, M., Boukerb, A. M., Racine, P.-J., Pichon, C., Kremser, C., Queiroz, A., Karsybayeva, M., Redziniak, G., Chevalier, S., & Feuilloley, M. G. J. (2019). Draft Genome Sequence of Lactobacillus crispatus Strain V4, Isolated from a Vaginal Swab from a Young Healthy Nonmenopausal Woman. Microbiology Resource Announcements, 8(38), 10.1128/mra.00856-19. https://doi.org/10.1128/mra.00856-19
10. Etges, W. J., de Oliveira, C. C., Rajpurohit, S., & Gibbs, A. G. (2016). Effects of temperature on transcriptome and cuticular hydrocarbon expression in ecologically differentiated populations of desert Drosophila. Ecology and Evolution, 7(2), 619–637. https://doi.org/10.1002/ece3.2653
11. Farmer, M., Rajasabhai, R., Tarpeh, W., Tyo, K., & Wells, G. (2023). Meta-omic profiling reveals ubiquity of genes encoding for the nitrogen-rich biopolymer cyanophycin in activated sludge microbiomes. Frontiers in Microbiology, 14, 1287491. https://doi.org/10.3389/fmicb.2023.1287491
12. Frisch, M. B., Castillo-Ramírez, S., Petit, R. A., Farley, M. M., Ray, S. M., Albrecht, V. S., Limbago, B. M., Hernandez, J., See, I., Satola, S. W., & Read, T. D. (2018). Invasive Methicillin-Resistant Staphylococcus aureus USA500 Strains from the U.S. Emerging Infections Program Constitute Three Geographically Distinct Lineages. mSphere, 3(3), 10.1128/msphere.00571-17. https://doi.org/10.1128/msphere.00571-17
13. Göker, M., Saunders, E., Lapidus, A., Nolan, M., Lucas, S., Hammon, N., Deshpande, S., Cheng, J.-F., Han, C., Tapia, R., Goodwin, L. A., Pitluck, S., Liolios, K., Mavromatis, K., Pagani, I., Ivanova, N., Mikhailova, N., Pati, A., Chen, A., … Klenk, H.-P. (2012). Genome sequence of the moderately thermophilic, amino-acid-degrading and sulfur-reducing bacterium Thermovirga lienii type strain (Cas60314T). Standards in Genomic Sciences, 6(2), 230–239. https://doi.org/10.4056/sigs.2726028
14. Hazards, E. P. on B., Koutsoumanis, K., Allende, A., Alvarez‐Ordóñez, A., Bolton, D., Bover‐Cid, S., Chemaly, M., Davies, R., De Cesare, A., Hilbert, F., Lindqvist, R., Nauta, M., Peixe, L., Ru, G., Simmons, M., Skandamis, P., Suffredini, E., Cocconcelli, P. S., Escámez, P. S. F., … Herman, L. (2022). Update of the list of QPS‐recommended microbiological agents intentionally added to food or feed as notified to EFSA 16: Suitability of taxonomic units notified to EFSA until March 2022. EFSA Journal, 20(7), e07408. https://doi.org/10.2903/j.efsa.2022.7408
15. Hermanns, K., Marklewitz, M., Zirkel, F., Overheul, G. J., Page, R. A., Loaiza, J. R., Drosten, C., van Rij, R. P., & Junglen, S. (2019). Agua Salud alphavirus defines a novel lineage of insect-specific alphaviruses discovered in the New World. The Journal of General Virology, 101(1), 96–104. https://doi.org/10.1099/jgv.0.001344
16. Howe, K. L., Zaugg, J., & Mason, O. U. (2024). Novel, active, and uncultured hydrocarbon-degrading microbes in the ocean. Applied and Environmental Microbiology, 90(9), e01224-24. https://doi.org/10.1128/aem.01224-24
17. Ichida, H., & Long, S. R. (2016). LDSS-P: an advanced algorithm to extract functional short motifs associated with coordinated gene expression. Nucleic Acids Research, 44(11), 5045–5053. https://doi.org/10.1093/nar/gkw435
18. Jagadeesan, Y., Meenakshisundaram, S., Boopathy, L. R. A., Mookandi, V. P. S., & Balaiah, A. (2019). Combinatorial approach for screening and assessment of multiple therapeutic enzymes from marine isolate Pseudomonas aeruginosa AR01. RSC Advances, 9(30), 16989–17001. https://doi.org/10.1039/c9ra02555c
19. Ji, H. J., Byun, E.-B., Chen, F., Ahn, K. B., Jung, H. K., Han, S. H., Lim, J. H., Won, Y., Moon, J. Y., Hur, J., & Seo, H. S. (2021). Radiation-Inactivated S. gallinarum Vaccine Provides a High Protective Immune Response by Activating Both Humoral and Cellular Immunity. Frontiers in Immunology, 12, 717556. https://doi.org/10.3389/fimmu.2021.717556
20. Lam, B., Kajderowicz, K. M., Keys, H. R., Roessler, J. M., Frenkel, E. M., Kirkland, A., Bisht, P., El-Brolosy, M. A., Jaenisch, R., Bell, G. W., Weissman, J. S., Griffith, E. C., & Hrvatin, S. (2024). Multi-species genome-wide CRISPR screens identify conserved suppressors of cold-induced cell death. bioRxiv, 2024.07.25.605098. https://doi.org/10.1101/2024.07.25.605098
21. Leavitt, J. C., Woodbury, B. M., Gilcrease, E. B., Bridges, C. M., Teschke, C. M., & Casjens, S. R. (2024). Bacteriophage P22 SieA-mediated superinfection exclusion. mBio, 15(2), e02169-23. https://doi.org/10.1128/mbio.02169-23
22. Lee, K., Badaya, S. K., Singh, R., & Lim, J. Y. (2020). Complete Genome Sequence of Gordonia sp. Strain JH63, Isolated from Human Skin. Microbiology Resource Announcements, 9(11), 10.1128/mra.00059-20. https://doi.org/10.1128/mra.00059-20
23. Li, X.-Q., & Du, D. (2012). Gene direction in living organisms. Scientific Reports, 2(1), 982. https://doi.org/10.1038/srep00982
24. Machado, H., & Gram, L. (2017). Comparative Genomics Reveals High Genomic Diversity in the Genus Photobacterium. Frontiers in Microbiology, 8, 1204. https://doi.org/10.3389/fmicb.2017.01204
25. Matsutani, M., Kawajiri, E., Yakushi, T., Adachi, O., & Matsushita, K. (2013). Draft Genome Sequence of Dihydroxyacetone-Producing Gluconobacter thailandicus Strain NBRC 3255. Genome Announcements, 1(2), e00118-13. https://doi.org/10.1128/genomea.00118-13
26. Moon, E.-K., Kong, H.-H., Hong, Y., Lee, H.-A., & Quan, F.-S. (2017). Identification and Characterization of Protein Arginine Methyltransferase 1 in Acanthamoeba castellanii. The Korean Journal of Parasitology, 55(2), 109–114. https://doi.org/10.3347/kjp.2017.55.2.109
27. Moreno, M., Marinotti, O., Krzywinski, J., Tadei, W. P., James, A. A., Achee, N. L., & Conn, J. E. (2010). Complete mtDNA genomes of Anopheles darlingi and an approach to anopheline divergence time. Malaria Journal, 9(1), 127. https://doi.org/10.1186/1475-2875-9-127
28. Mukherjee, S., Stamatis, D., Bertsch, J., Ovchinnikova, G., Verezemska, O., Isbandi, M., Thomas, A. D., Ali, R., Sharma, K., Kyrpides, N. C., & Reddy, T. B. K. (2016). Genomes OnLine Database (GOLD) v.6: Data updates and feature enhancements. Nucleic Acids Research, 45(D1), d446–d456. https://doi.org/10.1093/nar/gkw992
29. Nedashkovskaya, O., Balabanova, L., Otstavnykh, N., Zhukova, N., Detkova, E., Seitkalieva, A., Bystritskaya, E., Noskova, Y., Tekutyeva, L., & Isaeva, M. (2024). In-Depth Genome Characterization and Pan-Genome Analysis of Strain KMM 296, a Producer of Highly Active Alkaline Phosphatase; Proposal for the Reclassification of Cobetia litoralis and Cobetia pacifica as the Later Heterotypic Synonyms of Cobetia amphilecti and Cobetia marina, and Emended Description of the Species Cobetia amphilecti and Cobetia marina. Biomolecules, 14(2), 196. https://doi.org/10.3390/biom14020196
30. Otun, S. O., & Ntushelo, K. (2023). A data catalogue of alcohol dehydrogenases in the genome of a tissue macerating Pectobacterium brasiliense isolated from potato in South Africa. Data in Brief, 52, 109918. https://doi.org/10.1016/j.dib.2023.109918
31. Piña-Iturbe, A., Díaz-Gavidia, C., Álvarez, F. P., Barron-Montenegro, R., Álvarez-Espejo, D. M., García, P., Solís, D., Constenla-Albornoz, R., Toro, M., Olivares-Pacheco, J., Reyes-Jara, A., Meng, J., Bell, R. L., & Moreno-Switt, A. I. (2024). Genomic characterisation of the population structure and antibiotic resistance of Salmonella enterica serovar Infantis in Chile, 2009–2022. The Lancet Regional Health - Americas, 32, 100711. https://doi.org/10.1016/j.lana.2024.100711
32. Taniguti, L. M., Schaker, P. D. C., Benevenuto, J., Peters, L. P., Carvalho, G., Palhares, A., Quecine, M. C., Nunes, F. R. S., Kmit, M. C. P., Wai, A., Hausner, G., Aitken, K. S., Berkman, P. J., Fraser, J. A., Moolhuijzen, P. M., Coutinho, L. L., Creste, S., Vieira, M. L. C., Kitajima, J. P., & Monteiro-Vitorello, C. B. (2015). Complete Genome Sequence of Sporisorium scitamineum and Biotrophic Interaction Transcriptome with Sugarcane. PLOS ONE, 10(6), e0129318. https://doi.org/10.1371/journal.pone.0129318
33. Tijsse-Klasen, E., Pandak, N., Hengeveld, P., Takumi, K., Koopmans, M. P., & Sprong, H. (2013). Ability to cause erythema migrans differs between Borrelia burgdorferi sensu lato isolates. Parasites & Vectors, 6(1), 23. https://doi.org/10.1186/1756-3305-6-23
34. Tsai, T.-L., Su, C.-C., Hsieh, C.-C., Lin, C.-N., Chang, H.-W., Lo, C.-Y., Lin, C.-H., & Wu, H.-Y. (2018). Gene Variations in Cis-Acting Elements between the Taiwan and Prototype Strains of Porcine Epidemic Diarrhea Virus Alter Viral Gene Expression. Genes, 9(12), 591. https://doi.org/10.3390/genes9120591
35. Tsolis, R. M., Seshadri, R., Santos, R. L., Sangari, F. J., Lobo, J. M. G., de Jong, M. F., Ren, Q., Myers, G., Brinkac, L. M., Nelson, W. C., DeBoy, R. T., Angiuoli, S., Khouri, H., Dimitrov, G., Robinson, J. R., Mulligan, S., Walker, R. L., Elzer, P. E., Hassan, K. A., & Paulsen, I. T. (2009). Genome Degradation in Brucella ovis Corresponds with Narrowing of Its Host Range and Tissue Tropism. PLoS ONE, 4(5), e5519. https://doi.org/10.1371/journal.pone.0005519
36. Val-Calvo, J., Miguel-Arribas, A., Abia, D., Wu, L. J., & Meijer, W. J. J. (2021). pLS20 is the archetype of a new family of conjugative plasmids harboured by Bacillus species. NAR Genomics and Bioinformatics, 3(4), lqab096. https://doi.org/10.1093/nargab/lqab096
37. Vazquez-Anderson, J., Mihailovic, M. K., Baldridge, K. C., Reyes, K. G., Haning, K., Cho, S. H., Amador, P., Powell, W. B., & Contreras, L. M. (2017). Optimization of a novel biophysical model using large scale in vivo antisense hybridization data displays improved prediction capabilities of structurally accessible RNA regions. Nucleic Acids Research, 45(9), 5523–5538. https://doi.org/10.1093/nar/gkx115
38. Wang, J., Sun, S., Chen, Y., Chen, D., Sang, L., & Xie, X. (2022). Pathogenic and genomic characterisation of a rabbit sourced Pasteurella multocida serogroup F isolate s4. BMC Veterinary Research, 18(1), 288. https://doi.org/10.1186/s12917-022-03381-7
39. White, R. T., Legione, A. R., Taylor-Brown, A., Fernandez, C. M., Higgins, D. P., Timms, P., & Jelocnik, M. (2021). Completing the Genome Sequence of Chlamydia pecorum Strains MC/MarsBar and DBDeUG: New Insights into This Enigmatic Koala (Phascolarctos cinereus) Pathogen. Pathogens, 10(12), 1543. https://doi.org/10.3390/pathogens10121543
40. Wohlers, H., Zentgraf, L., van der Sande, L., & Holtmann, D. (2024). Metabolic engineering of Shewanella oneidensis to produce glutamate and itaconic acid. Applied Microbiology and Biotechnology, 108(1), 36. https://doi.org/10.1007/s00253-023-12879-5
41. Wu, Y., Wang, J., Xu, T., Liu, J., Yu, W., Lou, Q., Zhu, T., He, N., Ben, H., Hu, J., Götz, F., & Qu, D. (2012). The Two-Component Signal Transduction System ArlRS Regulates Staphylococcus epidermidis Biofilm Formation in an ica-Dependent Manner. PLoS ONE, 7(7), e40041. https://doi.org/10.1371/journal.pone.0040041
42. Xiong, L., Liu, S., Chen, S., Xiao, Y., Zhu, B., Gao, Y., Zhang, Y., Chen, B., Luo, J., Deng, Z., Chen, X., Wang, L., & Chen, S. (2019). A new type of DNA phosphorothioation-based antiviral system in archaea. Nature Communications, 10(1), 1688. https://doi.org/10.1038/s41467-019-09390-9



